### ISSUE TITLE
Conditional parameters - "map" argument in MakeADFun

### ISSUE THREAD

**Issue by merrillrudd** [James-Thorson/2016_Spatio-temporal_models#8]
Just wanted to share a little trick we haven't gone over yet in class that might be useful.

If you want to estimate a parameter for some models but not others, you can use the "map" argument in MakeADFun to fix a parameter at the starting value.

So say you have the function W = a*L^b and want to fix b=3 for some datasets but estimate it for others:
1) declare b as a parameter in your cpp file
2) provide a starting value of 3 for b in your "parameters" list to pass to MakeADFun
3) if you want to fix b=3 and not estimate it, then you can setup an additional list called "Map" that is a list of the parameters you want to fix at their starting values. You want to set those as factors of NA. From Jim, I use the code snippet:

Map <- NULL
Map[["b"]] <- NA
Map[["b"]] <- factor(Map[["b"]])

(There is probably a slicker way to write that but it works so I keep it. )

4) Obj <- MakeADFun(data=Data, parameters=Parameters, map=Map)

Another example is if you have a vector of parameters, but want to fix only some of them. I do this when I have a vector of fishing mortality rates I want to estimate as fixed effects but only have enough information to estimate the most recent few. So say I have set up the model to estimate 20 fishing mortality parameters but only want to estimate the last 10, I would:

1) define a vector "F_t" of length 20 in my Parameters list and set them all to a starting value.
2) Map[["F_t"]] <- 1:length(Parameters[["F_t"]])
    Map[["F_t"]][1:10] <- NA
    Map[["F_t"]] <- factor(Map[["F_t"]])

**Comment by colemonnahan** [comment]
Hi @merrillrudd Thanks for sharing this. Have you had any luck using it with random effects? It seems to work fine until the parameter that I unmap is also in the `random` argument. TMB just crashes at that point. If I estimate them as fixed effects then it runs fine, so it appears there's some kind of interaction between `map` and `random` that I don't understand.

@James-Thorson Any tips on that?

**Comment by James-Thorson** [comment]
What OS are you using?

**Comment by colemonnahan** [comment]
Windows 10

**Comment by merrillrudd** [comment]
Hmmm I have used it to "unmap" my random effects before and I believe it worked fine.

**Comment by James-Thorson** [comment]
Yeah, I've never seen that behavior, but TMB has updates periodically, so
maybe we have different versions.  I have definitely used "map" to turn off
parameters that are normally treated as random effects.  Are you willing to
attach a minimal example?

**Comment by colemonnahan** [comment]
I did a fresh install of TMB and it seems to be working fine now. Phew!

**Comment by RebeccaStubbs** [comment]
For what it's worth, I was also getting errors involving the map function-- TMB gets angry when you pass it things through the "Random" slot that are also masked, so you need to remove things from your list of "Random" variables if you're planning to map them out.
